azurelinuxagent/common/protocol/extensions_goal_state_from_vm_settings.py

```diff
@@ -267,10 +269,16 @@ def _parse_agent_manifests(self, vm_settings):
         for family in families:
             name = family["name"]
             version = family.get("version")
+            is_version_from_rsm = family.get("isVersionFromRSM")
+            is_vm_enabled_for_rsm_upgrades = family.get("isVMEnabledForRSMUpgrades")
             uris = family.get("uris")
             if uris is None:
                 uris = []
             agent_family = VMAgentFamily(name, version)
+            if is_version_from_rsm is not None:
+                agent_family.is_version_from_rsm = is_version_from_rsm
```

Comment:
Another related comment: Try to avoid modelling those flags as tri-state. Allowing None makes the code harder to write/review/maintain, since the consumers of those flags need to keep checking for None.

In other parts of the code, the flags are actually bools (they are never None) and that makes code hard to understand.
As an example, should agent_update_handler.py@409 be checking for None?

`if not self._is_version_from_rsm or self._version < self._daemon_version or self._version == CURRENT_VERSION`

(it shouldn't, but that is hard to understand without digging into the implementation in detail; it should not be that hard)

Reply:
The intention for tri-state modeling is agent_family should reflect what we see in the goal state. When we parse the goal state, if property missing (possible scenario) consider as None instead false because agent_update should know the state to compute latest state. If agent update last known as True, but in next goal state property missing, how do I know I should consider last known state (True) not the new state in the case of property missing. If I convert that into False, I wouldn't know if that's due to missing property or it's actual value. 

The state that we maintain in agent_update is two-state (True/False) but the agent_family model reflects the goal state(try-state).

Reply:
Yes, but that is very hard to understand without digging deep into the code. Shouldn't be this hard for a simple flag. There are better ways to model this.